test/types/symbol.js

```diff
@@ -0,0 +1,224 @@
+'use strict';
+
+// Load modules
+
+const Lab = require('lab');
+const Joi = require('../..');
+const Helper = require('../helper');
+
+
+// Declare internals
+
+const internals = {};
+
+
+// Test shortcuts
+
+const { describe, it, expect } = exports.lab = Lab.script();
+
+
+describe('symbol', () => {
+
+    it('cannot be called on its own', () => {
+
+        const symbol = Joi.symbol;
+        expect(() => symbol()).to.throw('Must be invoked on a Joi instance.');
+    });
+
+    it('should throw an exception if arguments were passed.', () => {
+
+        expect(
+            () => Joi.symbol('invalid argument.')
+        ).to.throw('Joi.symbol() does not allow arguments.');
+    });
+
+    describe('validate()', () => {
+
+        it('handles plain symbols', () => {
+
+            const symbols = [Symbol(1), Symbol(2)];
+            const rule = Joi.symbol();
+            Helper.validate(rule, [
+                [symbols[0], true, null, symbols[0]],
+                [symbols[1], true, null, symbols[1]],
+                [1, false, null, {
+                    message: '"value" must be a symbol',
+                    details: [{
+                        message: '"value" must be a symbol',
+                        path: [],
+                        type: 'symbol.base',
+                        context: { label: 'value', key: undefined }
+                    }]
+                }]
+            ]);
+        });
+
+        it('handles simple lookup', () => {
+
+            const symbols = [Symbol(1), Symbol(2)];
+            const otherSymbol = Symbol(1);
+            const rule = Joi.symbol().valid(symbols);
+            Helper.validate(rule, [
+                [symbols[0], true, null, symbols[0]],
+                [symbols[1], true, null, symbols[1]],
+                [otherSymbol, false, null, {
+                    message: '"value" must be one of [Symbol(1), Symbol(2)]',
+                    details: [{
+                        message: '"value" must be one of [Symbol(1), Symbol(2)]',
+                        path: [],
+                        type: 'any.allowOnly',
+                        context: { value: otherSymbol, label: 'value', valids: symbols, key: undefined }
+                    }]
+                }]
+            ]);
+        });
+
+        describe('map', () => {
+
+            it('converts keys to correct symbol', () => {
+
+                const symbols = [Symbol(1), Symbol(2), Symbol(3)];
+                const otherSymbol = Symbol(1);
+                const map = new Map([[1, symbols[0]], ['two', symbols[1]], [symbols[0], symbols[2]]]);
+                const rule = Joi.symbol().map(map);
+                Helper.validate(rule, [
+                    [1, true, null, symbols[0]],
+                    [symbols[0], true, null, symbols[0]],
```

Comment:
Why wouldn't it become `symbols[2]` ? Or else why accept symbols as key of a map ?

Reply:
Hmm, yeah I didn't write that test as I intended to.

However, it might makes sense to work as it does, since the implementation adds all map values to the `valid()` values, which is resolved before the mapping.

Reply:
Then I would deny symbols as keys, that doesn't make sense to me.

Reply:
Well, normally the mapping would work – it just doesn't in the case, since the symbol is also used as a value.

Anyway, I don't have a problem with removing it. We can always re-add later if it somehow makes sense.